v2/mongodb-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/mongodb/templates/MongoDbUtils.java

```diff
@@ -109,13 +135,89 @@ public static TableRow getTableSchema(HashMap<String, Object> parsedMap, String
+      String uri, String database, String collection, String udfGcsPath, String udfFunctionName)
+      throws IOException, ScriptException, NoSuchMethodException {
+    Document document = getMongoDbDocument(uri, database, collection);
+    List<TableFieldSchema> bigquerySchemaFields = new ArrayList<>();
+
+    ScriptEngineManager manager = new ScriptEngineManager();
+    ScriptEngine engine = manager.getEngineByName("JavaScript");
+    Collection<String> scripts = getScripts(udfGcsPath);
+
+    Invocable invocable = newInvocable(scripts);
+    if (invocable == null) {
+      throw new RuntimeException("No udf was loaded");
+    }
+
+    Object result = invocable.invokeFunction(udfFunctionName, document);
+    Document doc = (Document) result;
+    doc.forEach(
+        (key, value) -> {
+          bigquerySchemaFields.add(
+              new TableFieldSchema()
+                  .setName(key)
+                  .setType(getTableSchemaDataType(value.getClass().getName())));
+        });
+    bigquerySchemaFields.add(new TableFieldSchema().setName("timestamp").setType("TIMESTAMP"));
+    TableSchema bigquerySchema = new TableSchema().setFields(bigquerySchemaFields);
+    return bigquerySchema;
+  }
+
+  private static Collection<String> getScripts(String path) throws IOException {
+    MatchResult result = FileSystems.match(path);
+    checkArgument(
+        result.status() == Status.OK && !result.metadata().isEmpty(),
+        "Failed to match any files with the pattern: " + path);
+
+    List<String> scripts =
+        result.metadata().stream()
+            .filter(metadata -> metadata.resourceId().getFilename().endsWith(".js"))
+            .map(Metadata::resourceId)
+            .map(
+                resourceId -> {
+                  try (Reader reader =
+                      Channels.newReader(
+                          FileSystems.open(resourceId), StandardCharsets.UTF_8.name())) {
+                    return CharStreams.toString(reader);
+                  } catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                  }
+                })
+            .collect(Collectors.toList());
+    return scripts;
+  }
+
+  @Nullable
+  private static Invocable newInvocable(Collection<String> scripts) throws ScriptException {
+    ScriptEngineManager manager = new ScriptEngineManager();
+    ScriptEngine engine = manager.getEngineByName("JavaScript");
+
+    if (engine == null) {
+      List<String> availableEngines = new ArrayList<>();
+      for (ScriptEngineFactory factory : manager.getEngineFactories()) {
+        availableEngines.add(factory.getEngineName() + " " + factory.getEngineVersion());
+      }
+      throw new RuntimeException(
+          String.format("JavaScript engine not available. Found engines: %s.", availableEngines));
+    }
+
+    for (String script : scripts) {
+      engine.eval(script);
+    }
+
+    return (Invocable) engine;
+  }
```

Comment:
This duplicates from [JavascriptDocumentTransformer.java](https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/517/files#diff-c647c306f12e6001c60ef671fb78cd666a363e8412e5dd8e205d33f6e3cb4958) - can we consolidate/reuse?